examples/recording_service/leveldb-integration/cpp/LevelDbReader.cxx

```diff
@@ -0,0 +1,637 @@
+     * key comparator if we don't use the same one.
+     */
+    db_options.comparator = &key_comparator_;
+    leveldb::Status db_status =
+            leveldb::DB::Open(db_options, db_filename_, &db);
+    if (!db_status.ok() || db == nullptr) {
+        std::stringstream log_stream;
+        log_stream << "Failed to open user-data DB file" << std::endl;
+        log_stream << "    File name = " << db_filename_ << std::endl;
+        log_stream << "    LevelDB error = " << db_status.ToString()
+                << std::endl;
+        throw std::runtime_error(log_stream.str());
+    }
+    data_db_.reset(db);
+
+    /*
+     * Pre-allocate key buffer, used to serialize the UserDataKey type to be
+     * used as a key
+     */
+    const StructType& key_type = rti::topic::dynamic_type<UserDataKey>::get();
+    key_buffer_.resize(key_type.cdr_serialized_sample_max_size());
+
+    /*
+     * Setup the iterator. We need to advance to the next valid sample that fits
+     * within the specified time range.
+     */
+    db_iterator_ = std::unique_ptr<leveldb::Iterator>(
+            data_db_->NewIterator(leveldb::ReadOptions()));
+    for (db_iterator_->SeekToFirst(); db_iterator_->Valid(); ) {
+        UserDataKey current_key = slice_to_user_type<UserDataKey>(
+                db_iterator_->key(),
+                key_buffer_);
+        current_timestamp_ = current_key.reception_timestamp();
+        if (current_timestamp_ >= start_timestamp_
+                && current_timestamp_ <= end_timestamp_) {
+            break;
+        } else {
+            db_iterator_->Next();
+        }
+    }
+    /* The DB may be empty (no sample fulfills the conditions) */
+    if (!db_iterator_->Valid()) {
+        finished_ = true;
+    }
+    /*
+     * Pre-allocate an intelligent size for the CDR buffer to be used by the
+     * store operations. Get the maximum sample size from the type information.
+     * We protect the buffer size if the type is unbounded.
+     */
+    DDS_TypeCode *type = static_cast<DDS_TypeCode *>(
+            stream_info.type_info().type_representation());
+    DDS_ExceptionCode_t ex = DDS_NO_EXCEPTION_CODE;
+    if (DDS_TypeCode_is_unbounded(type, DDS_BOOLEAN_FALSE, &ex)
+            && ex != DDS_NO_EXCEPTION_CODE) {
+        value_buffer_.resize(8192);
+    } else {
+        DDS_UnsignedLong size = DDS_TypeCode_get_cdr_serialized_sample_max_size(
+                type,
+                DDS_AUTO_DATA_REPRESENTATION,
+                &ex);
+        if (ex != DDS_NO_EXCEPTION_CODE) {
+            std::stringstream log_stream;
+            log_stream << "Could not get serialized sample max size for stream"
+                    << std::endl;
+            log_stream << "    Stream name = " << stream_info.stream_name()
+                    << std::endl;
+            throw std::runtime_error(log_stream.str());
+        }
+        value_buffer_.resize((std::vector<char>::size_type) size);
```

Comment:
same comment than for the writer part. Perhaps is not worth doing any of this, in any case we're allocating for value.blob()

Reply:
We could go for the simplest case, although I prefer to avoid the allocations in the critical path for the bounded case.
But if you prefer it, I can change it, I'm not too attached to the solution.

Reply:
I'm fine with it.